File: 88. Merge Sorted Array/88. Merge Sorted Array.java
class Solution {
    public void merge(int[] nums1, int m, int[] nums2, int n) {
        // Initialize pointers
        int p1 = m - 1; // Pointer for nums1
        int p2 = n - 1; // Pointer for nums2
        int p = m + n - 1; // Pointer for writing

        // Merge arrays from the end
        while (p2 >= 0) {
            if (p1 >= 0 && nums1[p1] > nums2[p2]) {
                nums1[p] = nums1[p1];
                p1--;
            } else {
                nums1[p] = nums2[p2];
                p2--;
            }
            p--;
        }
    }

    // Example usage
    public static void main(String[] args) {
        Solution solution = new Solution();
        int[] nums1 = {1, 2, 3, 0, 0, 0};
        int m = 3;
        int[] nums2 = {2, 5, 6};
        int n = 3;
        
        solution.merge(nums1, m, nums2, n);
        
        System.out.print("Merged array: ");
        for (int num : nums1) {
            System.out.print(num + " ");
        }
        // Output: Merged array: 1 2 2 3 5 6
    }
}

File: 88. Merge Sorted Array/88. Merge Sorted Array.rs
impl Solution {
    pub fn merge(nums1: &mut Vec<i32>, m: i32, nums2: &mut Vec<i32>, n: i32) {
        let (mut m, mut n) = (m as usize, n as usize);
        let mut write_index = m + n - 1;

        while n > 0 {
            if m > 0 && nums1[m - 1] > nums2[n - 1] {
                nums1[write_index] = nums1[m - 1];
                m -= 1;
            } else {
                nums1[write_index] = nums2[n - 1];
                n -= 1;
            }
            write_index -= 1;
        }
    }
}
